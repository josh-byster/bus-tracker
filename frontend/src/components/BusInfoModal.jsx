import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Modal, ModalHeader, ModalBody, Button, ModalFooter } from 'reactstrap';
import { MAPBOX_API_KEY, getVehicleInfo, getStop } from '../util/api';
import '../styles/InfoModal.scss';

class BusInfoModal extends Component {
  defaultState = {
    nextStop: '',
    previousStop: '',
    imageExpanded: false,
    imgLoaded: false,
    lastUpdated: 0,
  };

  constructor(props) {
    super(props);
    this.state = this.defaultState;
  }

  componentDidUpdate(prevProps) {
    const { isOpen } = this.props;
    if (prevProps.isOpen === false && isOpen === true) {
      const mapURL = this.getMapURL();

      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({ mapURL, imgLoaded: false });
      this.getNextPrevStops();
    }
  }

  resetState = () => {
    this.setState();
  };

  getNameOfStop = async (stopID) => {
    if (stopID !== null) {
      const nextStop = await getStop(stopID);
      if (!nextStop.stops[0]) {
        return 'Unknown';
      }
      return nextStop.stops[0].stop_points.filter(
        (obj) => obj.stop_id === stopID
      )[0].stop_name;
    }
    return 'Unknown';
  };

  getNextPrevStops = async () => {
    const { busInfo } = this.props;
    let vehicles;
    try {
      ({ vehicles } = await getVehicleInfo(busInfo.vehicle_id));
    } catch (e) {
      this.setState({
        nextStop: 'Unknown',
        previousStop: 'Unknown',
        lastUpdated: 'N/A',
        mapURL: '',
      });
      return;
    }
    const nextStopID = vehicles[0].next_stop_id;
    const prevStopID = vehicles[0].previous_stop_id;
    const nextStopName = await this.getNameOfStop(nextStopID);
    const prevStopName = await this.getNameOfStop(prevStopID);

    const lastUpdatedDate = vehicles[0].last_updated;
    this.setState({
      nextStop: nextStopName,
      previousStop: prevStopName,
      lastUpdated: !Number.isNaN(new Date() - new Date(lastUpdatedDate))
        ? Math.round((new Date() - new Date(lastUpdatedDate)) / 1000)
        : 'Unknown',
    });
  };

  getMapURL = () => {
    const { busInfo, stopInfo } = this.props;
    if (busInfo.location !== undefined && stopInfo.stop_points !== undefined) {
      const long = busInfo.location.lon;
      const { lat } = busInfo.location;
      const stopLat = stopInfo.stop_points[0].stop_lat;
      const stopLong = stopInfo.stop_points[0].stop_lon;
      return `https://api.mapbox.com/styles/v1/mapbox/streets-v10/static/pin-s-bus+3498db(${long},${lat}),pin-s-information+e74c3c(${stopLong},${stopLat})/auto/400x400@2x?access_token=${MAPBOX_API_KEY}`;
    }
    return null;
  };

  toggleImageExpand = () =>
    this.setState((prevState) => {
      return {
        imageExpanded: !prevState.imageExpanded,
      };
    });

  render() {
    const { props, state } = this;
    // silencing the linter warnings
    const { isOpen, toggle } = this.props;
    return (
      <div>
        <Modal
          isOpen={isOpen}
          toggle={() => {
            this.setState(this.defaultState);
            toggle();
          }}
          className="info-modal"
        >
          <ModalHeader toggle={this.toggle} style={this.props.headerStyle}>
            {props.busInfo.headsign}
          </ModalHeader>
          <ModalBody>
            <img
              className={
                state.imageExpanded
                  ? 'img-fluid map-image full'
                  : 'img-fluid map-image'
              }
              alt="bus information"
              src={state.mapURL}
              style={state.imgLoaded ? {} : { visibility: 'hidden' }} // Toggle visibility based on the image loaded
              onLoad={() => this.setState({ imgLoaded: true })}
            />
            <p className="expand-text" onClick={this.toggleImageExpand}>
              {state.imageExpanded ? 'Make Smaller' : 'Expand'}{' '}
            </p>
            <p
              className="stop-info"
              style={
                state.imgLoaded || state.mapURL === ''
                  ? {}
                  : { visibility: 'hidden' }
              }
            >
              <b>Next Stop: </b>
              {state.nextStop}
              <br />
              <b>Previous Stop: </b>
              {state.previousStop}
              <br />
              <br />
              {state.lastUpdated !== 'Unknown' && (
                <i>Position last updated {state.lastUpdated} seconds ago</i>
              )}
            </p>
          </ModalBody>
          <ModalFooter>
            <Button
              color="secondary"
              onClick={() => {
                this.setState({});
                toggle();
              }}
            >
              Exit
            </Button>
          </ModalFooter>
        </Modal>
      </div>
    );
  }
}

BusInfoModal.propTypes = {
  toggle: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  busInfo: PropTypes.object.isRequired,
  stopInfo: PropTypes.object.isRequired,
  headerStyle: PropTypes.object,
};

export default BusInfoModal;

// <center><img class='img-fluid' style='border-radius:20px;' src="https://api.mapbox.com/styles/v1/mapbox/streets-v10/static/pin-s-bus+3498db(${long},${lat}),pin-s-information+e74c3c(${stop_long},${stop_lat})/auto/500x500@2x?access_token=${mapbox_api}"></img>
