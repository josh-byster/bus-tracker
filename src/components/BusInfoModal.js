import React, { Component } from "react";
import PropTypes from "prop-types";
import {
  Modal,
  ModalHeader,
  ModalBody,
  Button,
  ModalFooter,
  Card
} from "reactstrap";
import { MAPBOX_API_KEY, getVehicleInfo, getStop } from "../util/api.js";
import "../styles/InfoModal.scss";
class BusInfoModal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      mapURL: "",
      nextStop: "",
      previousStop: ""
    };
  }
  componentDidUpdate(prevProps) {
    if (prevProps.busInfo !== this.props.busInfo) {
      var mapURL = this.getMapURL();

      this.setState({ mapURL: mapURL, imgLoaded: false });
      this.getNextPrevStops();
      console.log(this.props.busInfo);
      console.log(this.props.stopInfo);
    }
  }

  getNextPrevStops = async () => {
    const { status, vehicles } = await getVehicleInfo(
      this.props.busInfo.vehicle_id
    );
    if (status.code === 200) {
      var nextStopID = vehicles[0].next_stop_id;
      var prevStopID = vehicles[0].previous_stop_id;

      if (nextStopID !== null) {
        var nextStop = await getStop(nextStopID);
        var nextStopName = nextStop.stops[0].stop_points.filter(obj => {
          return obj.stop_id === nextStopID;
        })[0].stop_name;
      } else {
        nextStopName = "Unknown";
      }

      if (prevStopID !== null) {
        var prevStop = await getStop(prevStopID);
        var prevStopName = prevStop.stops[0].stop_points.filter(obj => {
          return obj.stop_id === prevStopID;
        })[0].stop_name;
      } else {
        prevStopName = "Unknown";
      }

      this.setState({ nextStop: nextStopName, previousStop: prevStopName });
    }
  };

  getMapURL = () => {
    if (
      this.props.busInfo.location !== undefined &&
      this.props.stopInfo.stop_points !== undefined
    ) {
      var long = this.props.busInfo.location.lon;
      var lat = this.props.busInfo.location.lat;
      var stop_lat = this.props.stopInfo.stop_points[0].stop_lat;
      var stop_long = this.props.stopInfo.stop_points[0].stop_lon;
      var mapbox_api = MAPBOX_API_KEY;
      return `https://api.mapbox.com/styles/v1/mapbox/streets-v10/static/pin-s-bus+3498db(${long},${lat}),pin-s-information+e74c3c(${stop_long},${stop_lat})/auto/400x400@2x?access_token=${mapbox_api}`;
    }
  };

  render() {
    return (
      <div>
        <Modal
          isOpen={this.props.isOpen}
          toggle={e => this.props.toggle()}
          className="info-modal"
        >
          <ModalHeader toggle={this.toggle}>
            {this.props.busInfo.headsign}
          </ModalHeader>
          <ModalBody>
            <img
              className="img-fluid map-image"
              alt="bus information"
              src={this.state.mapURL}
              style={this.state.imgLoaded ? {} : { visibility: "hidden" }} // Toggle visibility based on the image loaded
              onLoad={() => this.setState({ imgLoaded: true })}
            />

            <p className="stop-info">
              <b>Next Stop:</b> {this.state.nextStop}
              <br />
              <b>Previous Stop</b> {this.state.previousStop}
            </p>
          </ModalBody>
          <ModalFooter>
            <Button color="secondary" onClick={e => this.props.toggle()}>
              Exit
            </Button>
          </ModalFooter>
        </Modal>
      </div>
    );
  }
}

BusInfoModal.propTypes = {
  toggle: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  busInfo: PropTypes.object.isRequired,
  stopInfo: PropTypes.object.isRequired
};

export default BusInfoModal;

// <center><img class='img-fluid' style='border-radius:20px;' src="https://api.mapbox.com/styles/v1/mapbox/streets-v10/static/pin-s-bus+3498db(${long},${lat}),pin-s-information+e74c3c(${stop_long},${stop_lat})/auto/500x500@2x?access_token=${mapbox_api}"></img>
