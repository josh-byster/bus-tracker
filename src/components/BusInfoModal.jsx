import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Modal, ModalHeader, ModalBody, Button, ModalFooter } from 'reactstrap';
import { MAPBOX_API_KEY, getVehicleInfo, getStop } from '../util/api';
import '../styles/InfoModal.scss';

class BusInfoModal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      mapURL: '',
      nextStop: '',
      previousStop: ''
    };
  }

  componentDidUpdate(prevProps) {
    const { busInfo } = this.props;
    if (prevProps.busInfo !== busInfo) {
      const mapURL = this.getMapURL();

      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({ mapURL, imgLoaded: false });
      this.getNextPrevStops();
    }
  }

  getNextPrevStops = async () => {
    const { busInfo } = this.props;
    const { status, vehicles } = await getVehicleInfo(busInfo.vehicle_id);
    let nextStopName;
    let prevStopName;
    if (status.code === 200) {
      const nextStopID = vehicles[0].next_stop_id;
      const prevStopID = vehicles[0].previous_stop_id;

      if (nextStopID !== null) {
        const nextStop = await getStop(nextStopID);
        nextStopName = nextStop.stops[0].stop_points.filter(
          obj => obj.stop_id === nextStopID
        )[0].stop_name;
      } else {
        nextStopName = 'Unknown';
      }

      if (prevStopID !== null) {
        const prevStop = await getStop(prevStopID);
        prevStopName = prevStop.stops[0].stop_points.filter(
          obj => obj.stop_id === prevStopID
        )[0].stop_name;
      } else {
        prevStopName = 'Unknown';
      }

      this.setState({
        nextStop: nextStopName,
        previousStop: prevStopName,
        last_updated: vehicles[0].last_updated
      });
    }
  };

  getMapURL = () => {
    const { busInfo, stopInfo } = this.props;
    if (busInfo.location !== undefined && stopInfo.stop_points !== undefined) {
      const long = busInfo.location.lon;
      const { lat } = busInfo.location;
      const stopLat = stopInfo.stop_points[0].stop_lat;
      const stopLong = stopInfo.stop_points[0].stop_lon;
      return `https://api.mapbox.com/styles/v1/mapbox/streets-v10/static/pin-s-bus+3498db(${long},${lat}),pin-s-information+e74c3c(${stopLong},${stopLat})/auto/400x400@2x?access_token=${MAPBOX_API_KEY}`;
    }
    return null;
  };

  render() {
    const { props, state } = this;
    // silencing the linter warnings
    const { isOpen, toggle } = this.props;
    return (
      <div>
        <Modal isOpen={isOpen} toggle={() => toggle()} className="info-modal">
          <ModalHeader toggle={this.toggle}>
            {props.busInfo.headsign}
          </ModalHeader>
          <ModalBody>
            <img
              className="img-fluid map-image"
              alt="bus information"
              src={state.mapURL}
              style={state.imgLoaded ? {} : { visibility: 'hidden' }} // Toggle visibility based on the image loaded
              onLoad={() => this.setState({ imgLoaded: true })}
            />

            <p
              className="stop-info"
              style={state.imgLoaded ? {} : { visibility: 'hidden' }}
            >
              <b>Next Stop: </b>
              {state.nextStop}
              <br />
              <b>Previous Stop: </b>
              {state.previousStop}
              <br />
              <br />
              <i>
                Position last updated{' '}
                {Math.round((new Date() - new Date(state.last_updated)) / 1000)}{' '}
                seconds ago
              </i>
            </p>
          </ModalBody>
          <ModalFooter>
            <Button color="secondary" onClick={() => props.toggle()}>
              Exit
            </Button>
          </ModalFooter>
        </Modal>
      </div>
    );
  }
}

BusInfoModal.propTypes = {
  toggle: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  busInfo: PropTypes.object.isRequired,
  stopInfo: PropTypes.object.isRequired
};

export default BusInfoModal;

// <center><img class='img-fluid' style='border-radius:20px;' src="https://api.mapbox.com/styles/v1/mapbox/streets-v10/static/pin-s-bus+3498db(${long},${lat}),pin-s-information+e74c3c(${stop_long},${stop_lat})/auto/500x500@2x?access_token=${mapbox_api}"></img>
