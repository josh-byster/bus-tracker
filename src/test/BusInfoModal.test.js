import React from "react";
import expect from "expect";
import { shallow, mount } from "enzyme";
import BusInfoModal from "../components/BusInfoModal";
import EnzymeToJson from "enzyme-to-json";
import departures from "./mock_departures";
import renderer from "react-test-renderer";
// setup file
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

describe("Bus Info Modal Component", () => {
  const elem = mount(
    <BusInfoModal
      toggle={function() {}}
      isOpen={true}
      stopInfo={{}}
      busInfo={{ headsign: "12W Teal" }}
    />
  );
  elem.setState({
    mapURL: "http://example.com/",
    nextStop: "Test 1",
    previousStop: "Test 2"
  });

  it("renders the proper modal", () => {
    expect(EnzymeToJson(elem)).toMatchSnapshot();
  });
});
