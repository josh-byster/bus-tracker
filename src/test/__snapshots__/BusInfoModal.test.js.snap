// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bus Info Modal Component renders the proper modal 1`] = `
<BusInfoModal
  busInfo={
    Object {
      "headsign": "12W Teal",
    }
  }
  isOpen={true}
  stopInfo={Object {}}
  toggle={[Function]}
>
  <div>
    <Modal
      autoFocus={true}
      backdrop={true}
      backdropTransition={
        Object {
          "mountOnEnter": true,
          "timeout": 150,
        }
      }
      centered={false}
      className="info-modal"
      fade={true}
      isOpen={true}
      keyboard={true}
      modalTransition={
        Object {
          "timeout": 300,
        }
      }
      onClosed={[Function]}
      onOpened={[Function]}
      role="dialog"
      toggle={[Function]}
      zIndex={1050}
    >
      <Portal
        node={
          <div
            style="position: relative; z-index: 1050;"
            tabindex="-1"
          >
            <div
              class=""
            >
              <div
                class="modal fade"
                role="dialog"
                style="display: block;"
                tabindex="-1"
              >
                <div
                  class="modal-dialog info-modal"
                  role="document"
                >
                  <div
                    class="modal-content"
                  >
                    <div
                      class="modal-header"
                    >
                      <h5
                        class="modal-title"
                      >
                        12W Teal
                      </h5>
                    </div>
                    <div
                      class="modal-body"
                    >
                      <img
                        alt="bus information"
                        class="img-fluid map-image"
                        src="http://example.com/"
                        style="visibility: hidden;"
                      />
                      <p
                        class="stop-info"
                        style="visibility: hidden;"
                      >
                        <b>
                          Next Stop:
                        </b>
                         
                        Test 1
                        <br />
                        <b>
                          Previous Stop
                        </b>
                         
                        Test 2
                      </p>
                    </div>
                    <div
                      class="modal-footer"
                    >
                      <button
                        class="btn btn-secondary"
                        type="button"
                      >
                        Exit
                      </button>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="modal-backdrop fade"
              />
            </div>
          </div>
        }
      >
        <Portal
          containerInfo={
            <div
              style="position: relative; z-index: 1050;"
              tabindex="-1"
            >
              <div
                class=""
              >
                <div
                  class="modal fade"
                  role="dialog"
                  style="display: block;"
                  tabindex="-1"
                >
                  <div
                    class="modal-dialog info-modal"
                    role="document"
                  >
                    <div
                      class="modal-content"
                    >
                      <div
                        class="modal-header"
                      >
                        <h5
                          class="modal-title"
                        >
                          12W Teal
                        </h5>
                      </div>
                      <div
                        class="modal-body"
                      >
                        <img
                          alt="bus information"
                          class="img-fluid map-image"
                          src="http://example.com/"
                          style="visibility: hidden;"
                        />
                        <p
                          class="stop-info"
                          style="visibility: hidden;"
                        >
                          <b>
                            Next Stop:
                          </b>
                           
                          Test 1
                          <br />
                          <b>
                            Previous Stop
                          </b>
                           
                          Test 2
                        </p>
                      </div>
                      <div
                        class="modal-footer"
                      >
                        <button
                          class="btn btn-secondary"
                          type="button"
                        >
                          Exit
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="modal-backdrop fade"
                />
              </div>
            </div>
          }
        >
          <div
            className=""
          >
            <Fade
              appear={true}
              baseClass="fade"
              baseClassActive="show"
              className="modal"
              enter={true}
              exit={true}
              in={true}
              mountOnEnter={false}
              onClick={[Function]}
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              role="dialog"
              style={
                Object {
                  "display": "block",
                }
              }
              tabIndex="-1"
              tag="div"
              timeout={300}
              unmountOnExit={false}
            >
              <Transition
                appear={true}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={false}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={300}
                unmountOnExit={false}
              >
                <div
                  className="modal fade"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  role="dialog"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                  tabIndex="-1"
                >
                  <div
                    className="modal-dialog info-modal"
                    role="document"
                  >
                    <div
                      className="modal-content"
                    >
                      <ModalHeader
                        charCode={215}
                        closeAriaLabel="Close"
                        tag="h5"
                        wrapTag="div"
                      >
                        <div
                          className="modal-header"
                        >
                          <h5
                            className="modal-title"
                          >
                            12W Teal
                          </h5>
                        </div>
                      </ModalHeader>
                      <ModalBody
                        tag="div"
                      >
                        <div
                          className="modal-body"
                        >
                          <img
                            alt="bus information"
                            className="img-fluid map-image"
                            onLoad={[Function]}
                            src="http://example.com/"
                            style={
                              Object {
                                "visibility": "hidden",
                              }
                            }
                          />
                          <p
                            className="stop-info"
                            style={
                              Object {
                                "visibility": "hidden",
                              }
                            }
                          >
                            <b>
                              Next Stop:
                            </b>
                             
                            Test 1
                            <br />
                            <b>
                              Previous Stop
                            </b>
                             
                            Test 2
                          </p>
                        </div>
                      </ModalBody>
                      <ModalFooter
                        tag="div"
                      >
                        <div
                          className="modal-footer"
                        >
                          <Button
                            color="secondary"
                            onClick={[Function]}
                            tag="button"
                          >
                            <button
                              aria-label={null}
                              className="btn btn-secondary"
                              onClick={[Function]}
                              type="button"
                            >
                              Exit
                            </button>
                          </Button>
                        </div>
                      </ModalFooter>
                    </div>
                  </div>
                </div>
              </Transition>
            </Fade>
            <Fade
              appear={true}
              baseClass="fade"
              baseClassActive="show"
              className="modal-backdrop"
              enter={true}
              exit={true}
              in={true}
              mountOnEnter={true}
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              tag="div"
              timeout={150}
              unmountOnExit={false}
            >
              <Transition
                appear={true}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={true}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={150}
                unmountOnExit={false}
              >
                <div
                  className="modal-backdrop fade"
                />
              </Transition>
            </Fade>
          </div>
        </Portal>
      </Portal>
    </Modal>
  </div>
</BusInfoModal>
`;
